Fix for numpy arrays when the size 1 arrays were treated as scalars

mijo2 opened the pull request:
#### References to other Issues or PRs
Revives and closes #14720 
Fixes #14706

#### Brief description of what is fixed or changed

--> Added a bug fix in sympify whereby the numpy arrays of size 1 won't be treated as scalars

#### Other comments

Example:
```
>>> from sympy import *
>>> from sympy.abc import *
>>> import numpy
>>> z = np.zeros((1,1))
```

Before the fix:
```
>>> f = x + z
x
```

After the fix:
```
>>> f = x + z
array([[x]], dtype=object)
```

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* core
  *  sympify no longer converts numpy arrays with size to scalars
<!-- END RELEASE NOTES -->

NeilGirdhar committed:
Fixed a bug whereby numpy arrays of size one were treated as scalars.

NeilGirdhar committed:
Added a test for issue 14706.

NeilGirdhar committed:
Updated tests.

NeilGirdhar committed:
Updated tests.

mijo2 committed:
Merge branch 'master' of https://github.com/NeilGirdhar/sympy into 14720_fix_for_numpy_sympy

core: Changed the last test in test_issue_14720

asmeurer commented:
We all the comments from the original PR addressed here?

mijo2 commented:
@asmeurer I think so. The main thing was changing the last test to a Sympify error check. But I will add more test cases for Sympify if required.

czgdp1807 commented:
More tests are needed, diff coverage is too low.

mijo2 committed:
core: More test cases for sympify

Particularly related to the new changes, new tests were added to
increase the coverage.

smichr commented on sympy/core/tests/test_sympify.py:
```diff
@@ -695,3 +696,39 @@ def test_issue_16759():
+    if not numpy:
+        skip("numpy not installed.")
+
+    from sympy import symbols
```
```suggestion
```

smichr commented on sympy/core/tests/test_sympify.py:
```diff
@@ -695,3 +696,39 @@ def test_issue_16759():
+
+    from sympy import symbols
+
+    x = symbols('x')
```
```suggestion
    x = Symbol('x')
```

mijo2 committed:
core: removed unnecessary imports and definitions from test_sympify.py

mijo2 commented:
@asmeurer @smichr I think this is ready to be merged now.

smichr commented:
A comment on the orginal issue was

> It looks like the numpy people realized their mistake (numpy/numpy#10404) and are trying to deprecate it by the way, so in the long run this may be fixed on their end, and a link to the numpy issue might be good to add in a comment in the source code.

If the issue is fixed then can these changes be dropped or will these changes still be necessary. If the former then what we need is a test which passes now but will fail when numpy fixes the issue. A string in the function would indicate that this PR should be reverted at that time, e.g.
```python
def test_numpy_issue_10404():
    # if this fails then the PR in which this test was added
    # can be reverted (or whatever the instructions are)
    assert something that passes
```

mijo2 commented:
Ok @smichr I will take a look and add these kind of tests along with comment about what to do and which commit to PR.

mijo2 committed:
core: Updated some tests for later recognition of the cases when
numpy fixes their side

Whenever numpy fixes their side of the problem, the tests that are
added in this commit may fail. If they do, then appropriate comments
are added for someone in future to take up the problem and fix it.

mijo2 commented:
@smichr I have added the kind of tests along with appropriate comments for someone in future to take up the issue  if ever that happens.

mijo2 committed:
core: Minor fix

Minor bug in the code was fixed.

smichr commented:
OK, thanks for this fix.
